Revise the solution of rekeying the initial Child SA

William-PanWei opened the pull request.

William-PanWei committed:
change the solution of rekeying the initial Child SA

paulwouters commented on draft-ietf-ipsecme-ikev2-sa-ts-payloads-opt.md:
```diff
@@ -194,16 +194,10 @@ HDR, SK {N(REKEY_SA,oldSPI), N(OPTIMIZED_REKEY,newSPIi),
 
-For the initial Child SA that was negotiated as part of an initial IKE exchange (e.g., IKE_AUTH), at the time of its creation the parameters of PFS and KE method for Child SAs are not negotiated. Therefore, the KE method for the initial IKE SA should also be recognized as the one for this initial Child SA.
+For the initial Child SA that was negotiated as part of an initial IKE exchange (e.g., IKE_AUTH), at the time of its creation the parameters of PFS and KE method for Child SAs are not negotiated.
+Therefore, when rekeying the initial Child SA for the first time, a regular rekey MUST be used for the negotiation of these parameters. Then, the next rekey can use the optimized way.
```
i don't want to unconditionally say that, as i am still hoping to kill the regular rekey.

William-PanWei replied:
I've revised the text again. Now it highlights that the optimized rekey can be used if draft-child-pfs-info is used. You can have a check of the new text.

William-PanWei committed:
revise the text of rekeying the initial Child SA

Add a dedicated section for rekeying the initial Child SA.
Move the reference of draft-child-pfs-info to normative.
Revise the text to say the optimized rekey should be used when draft-child-pfs-info is used, and the regular rekey should be used when draft-child-pfs-info is not used

smyslov commented on draft-ietf-ipsecme-ikev2-sa-ts-payloads-opt.md:
```diff
@@ -292,7 +292,7 @@ OPTIMIZED_REKEY                          TBD2
 Some implementations allow sending rekey messages with a different set of Traffic Selectors or cryptographic parameters in response to a configuration update. IKEv2 [RFC7296] states this "SHOULD NOT" be done. But if there is a configuration change that changes the Traffic Selectors, cryptographic parameters, or other properties of the SA, the regular rekey should be used to make the configuration ...
 
-Two peers' PFS policy and KE method configurations MUST be the same, otherwise, the rekey of the Child SA created in the IKE_AUTH exchange would fail. This issue is also discussed in detail in {{-child-pfs}}. If the KE method for Child SAs is negotiated during the creation of the initial Child SA via the mechanism like {{-child-pfs}}, this KE method MUST be inherited when using the optimized meth ...
+Two peers' PFS policy and KE method configurations MUST be the same, otherwise, the rekey (regardless of regular or optimized way) of the initial Child SA created in the IKE_AUTH exchange would fail. This issue is also discussed in detail in {{-child-pfs}}. 
```
Why configurations "MUST be the same"? They MUST contain mutually accepatable KE methods, but they need not be the same.

William-PanWei replied:
```suggestion
Two peers MUST have the same PFS policy and contain mutally acceptable KE method(s), otherwise, the rekey (regardless of regular or optimized way) of the initial Child SA created in the IKE_AUTH exchange would fail. This issue is also discussed in detail in {{-child-pfs}}. 
```

paulwouters replied:
I think "the same" refers to "the same list or options", so I think it is okay.
If they were truly not the same list, you would get different behaviour based on being initiator or responder, which would be a fairly bogus policy to allow.

smyslov commented on draft-ietf-ipsecme-ikev2-sa-ts-payloads-opt.md:
```diff
@@ -194,16 +194,16 @@ HDR, SK {N(REKEY_SA,oldSPI), N(OPTIMIZED_REKEY,newSPIi),
-For the initial Child SA that was negotiated as part of an initial IKE exchange (e.g., IKE_AUTH), at the time of its creation the parameters of PFS and KE method for Child SAs are not negotiated. Therefore, the KE method for the initial IKE SA should also be recognized as the one for this initial Child SA.
-
-Two peers must have the same configurations for the parameters of PFS and KE method for Child SAs.
+If a responder fails to process the optimized rekey request, e.g., there is a change of the configuration, it MUST return an error as the notification of type NO_PROPOSAL_CHOSEN. After receiving the error response of the optimized rekey, the initiator can retry a regular rekey.
```
This is a bit vague. What if responder fails to process for some other reason, e.g. it cannot create more SAs? I think we should be more specific: "If a responder fails to process the optimized rekey request because for some reason it cannot re-use SA parameters for the SA being re-reyed (e.g., there is a change in the responder's configuration), ..." or something like this.

William-PanWei replied:
```suggestion
If a responder fails to process the optimized rekey request because for some reasons it cannot re-use SA parameters for the SA being rekeyed (e.g., there is a change in the responder's configuration), it MUST return an error as the notification of type NO_PROPOSAL_CHOSEN. After receiving the error response of the optimized rekey, the initiator can retry a regular rekey.
```

smyslov commented on draft-ietf-ipsecme-ikev2-sa-ts-payloads-opt.md:
```diff
@@ -194,16 +194,16 @@ HDR, SK {N(REKEY_SA,oldSPI), N(OPTIMIZED_REKEY,newSPIi),
 
-If the responder fails to process the optimized rekey request, e.g., receiving a request with a non-allowed PFS proposal, it MUST return an error as the notification of type NO_PROPOSAL_CHOSEN. After receiving the error response of the optimized rekey, the initiator can retry a regular rekey.
+If both peers support and use {{-child-pfs}}, the PFS policy and KE method(s) for the initial Child SA is known during its creation.
+Therefore, when rekeying the initial Child SA for the first time, the optimized way SHOULD be used.
```
I woud add: "Therefore, in this situation, ..." to emhasizi that both peer should know KE method(s).

William-PanWei replied:
```suggestion
Therefore, in this situation, when rekeying the initial Child SA for the first time, the optimized way SHOULD be used.
```

William-PanWei committed:
Apply suggestions from code review
